File: ts/tests/properties.test.ts
/**
 * Coverage for three estimator properties:
 *   - Sample-path bounds re-attach Ratio/Disparity units, while the RAW
 *     (native-array) bounds stay unitless (Number). Center/Spread/Shift Sample
 *     units propagate from x / finer(x, y).
 *   - The RAW (double-misrate) bounds API rejects out-of-[0,1]/NaN misrate with
 *     a domain/misrate AssumptionError.
 *   - The n==2 center midpoint (0.5*a + 0.5*b) is exactly order-symmetric:
 *     center([-5,-1.8]) bit-equals center([-1.8,-5]).
 */

import {
  center,
  centerBounds,
  spreadBounds,
  shiftBounds,
  ratioBounds,
  disparityBounds,
} from '../src/estimators';
import { AssumptionError } from '../src/assumptions';
import { MeasurementUnit } from '../src/measurement-unit';
import { Sample } from '../src/sample';

// A non-default unit to prove center/spread/shift propagate x's (finer) unit
// rather than hard-coding Number.
const SEC = new MeasurementUnit('sec', 'Time', 's', 'Second', 1);

describe('bounds unit re-attachment', () => {
  // Larger samples so the shuffle-based spread/disparity bounds clear their
  // min-achievable-misrate floors at the misrate below.
  const x = [2, 4, 6, 8, 10, 12, 14, 16, 18, 20];
  const y = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
  const sx = Sample.of(x);
  const sy = Sample.of(y);
  const misrate = 0.5;

  it('ratioBounds(Sample, Sample).unit === Ratio', () => {
    expect(ratioBounds(sx, sy, misrate).unit).toBe(MeasurementUnit.RATIO);
  });

  it('disparityBounds(Sample, Sample).unit === Disparity', () => {
    expect(disparityBounds(sx, sy, misrate).unit).toBe(MeasurementUnit.DISPARITY);
  });

  it('RAW ratioBounds is unitless (Number)', () => {
    expect(ratioBounds(x, y, misrate).unit).toBe(MeasurementUnit.NUMBER);
  });

  it('RAW disparityBounds is unitless (Number)', () => {
    expect(disparityBounds(x, y, misrate).unit).toBe(MeasurementUnit.NUMBER);
  });

  it('centerBounds(Sample).unit propagates x.unit', () => {
    const sxSec = Sample.withUnit(x, SEC);
    expect(centerBounds(sxSec, misrate).unit).toBe(SEC);
  });

  it('spreadBounds(Sample).unit propagates x.unit', () => {
    const sxSec = Sample.withUnit(x, SEC);
    expect(spreadBounds(sxSec, misrate).unit).toBe(SEC);
  });

  it('shiftBounds(Sample, Sample).unit propagates finer(x, y)', () => {
    const sxSec = Sample.withUnit(x, SEC);
    const sySec = Sample.withUnit(y, SEC);
    expect(shiftBounds(sxSec, sySec, misrate).unit).toBe(SEC);
  });

  it('RAW centerBounds / spreadBounds are unitless (Number)', () => {
    expect(centerBounds(x, misrate).unit).toBe(MeasurementUnit.NUMBER);
    expect(spreadBounds(x, misrate).unit).toBe(MeasurementUnit.NUMBER);
  });

  it('RAW shiftBounds is unitless (Number)', () => {
    expect(shiftBounds(x, y, misrate).unit).toBe(MeasurementUnit.NUMBER);
  });
});

describe('RAW bounds reject out-of-domain misrate', () => {
  const x = [1, 2, 3, 4, 5];
  const y = [2, 3, 4, 5, 6];

  function expectDomainMisrate(fn: () => unknown): void {
    let thrown: AssumptionError | null = null;
    try {
      fn();
    } catch (e) {
      if (e instanceof AssumptionError) thrown = e;
    }
    expect(thrown).not.toBeNull();
    expect(thrown!.violation!.id).toBe('domain');
    expect(thrown!.violation!.subject).toBe('misrate');
  }

  // One-sample (centerBounds) via the RAW/double-misrate path.
  it('centerBounds rejects misrate = 2.0', () => {
    expectDomainMisrate(() => centerBounds(x, 2.0));
  });

  it('centerBounds rejects misrate = -0.1', () => {
    expectDomainMisrate(() => centerBounds(x, -0.1));
  });

  it('centerBounds rejects misrate = NaN', () => {
    expectDomainMisrate(() => centerBounds(x, NaN));
  });

  // Two-sample (shiftBounds) via the RAW/double-misrate path.
  it('shiftBounds rejects misrate = 2.0', () => {
    expectDomainMisrate(() => shiftBounds(x, y, 2.0));
  });

  it('shiftBounds rejects misrate = -0.1', () => {
    expectDomainMisrate(() => shiftBounds(x, y, -0.1));
  });

  it('shiftBounds rejects misrate = NaN', () => {
    expectDomainMisrate(() => shiftBounds(x, y, NaN));
  });
});

describe('n==2 center midpoint is order-symmetric', () => {
  // centerImpl handles n==2 with an early return that runs BEFORE the
  // normalizing sort, so both argument orders feed the 0.5*a + 0.5*b midpoint
  // as-is. That formula is exactly order-symmetric: each operand is halved
  // independently, and the final addition commutes. An alternative midpoint
  // such as a + (b - a) * 0.5 is NOT: it yields -3.4 for [-5,-1.8] but
  // -3.4000000000000004 for [-1.8,-5]. Hence exact equality below, not
  // approximate.
  it('center([-5,-1.8]) bit-equals center([-1.8,-5])', () => {
    const forward = center([-5.0, -1.8]);
    const reversed = center([-1.8, -5.0]);
    expect(reversed).toBe(forward);
    expect(forward).toBe(-3.4);
  });
});

File: go/properties_test.go
package pragmastat

import (
	"math"
	"testing"
)

// Property tests that have no shared cross-language fixture:
// bounds unit re-attachment, the raw-path misrate domain branch, and the n==2
// center midpoint order-symmetry guard.

// testSecond is a custom (non-Number) unit used to prove that center/spread/
// shift bounds re-attach the *sample's* unit rather than the raw NumberUnit.
var testSecond = &MeasurementUnit{
	ID:           "second",
	Family:       "Time",
	Abbreviation: "s",
	FullName:     "Second",
	BaseUnits:    1,
}

// --- unit re-attachment ---------------------------------------------------
//
// The Sample-path ratio/disparity bounds re-attach Ratio/Disparity units, while
// the raw (native-slice) bounds API returns a unitless Number unit. center/
// spread/shift Sample bounds propagate the sample's unit (x / finer(x,y)).

func TestBoundsUnitReattachment(t *testing.T) {
	// 0.2 clears the disparity-bounds minimum (minShift+minAvg) for n=m=10
	// while remaining valid for center/spread/shift/ratio bounds.
	const misrate = 0.2
	x := []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	y := []float64{2, 4, 6, 8, 10, 12, 14, 16, 18, 20}

	sx, err := NewSampleWithUnit(x, testSecond)
	if err != nil {
		t.Fatalf("NewSampleWithUnit x: %v", err)
	}
	sy, err := NewSampleWithUnit(y, testSecond)
	if err != nil {
		t.Fatalf("NewSampleWithUnit y: %v", err)
	}

	// Sample path: ratio/disparity bounds carry the dedicated units.
	rb, err := sx.RatioBounds(sy, misrate)
	if err != nil {
		t.Fatalf("Sample.RatioBounds: %v", err)
	}
	if rb.Unit != RatioUnit {
		t.Errorf("Sample RatioBounds unit = %v, want RatioUnit", rb.Unit)
	}

	db, err := sx.DisparityBounds(sy, misrate)
	if err != nil {
		t.Fatalf("Sample.DisparityBounds: %v", err)
	}
	if db.Unit != DisparityUnit {
		t.Errorf("Sample DisparityBounds unit = %v, want DisparityUnit", db.Unit)
	}

	// Raw path: ratio/disparity bounds are UNITLESS (NumberUnit).
	rawRb, err := RatioBounds(x, y, misrate, false)
	if err != nil {
		t.Fatalf("RatioBounds raw: %v", err)
	}
	if rawRb.Unit != NumberUnit {
		t.Errorf("raw RatioBounds unit = %v, want NumberUnit", rawRb.Unit)
	}

	rawDb, err := DisparityBounds(x, y, misrate, false)
	if err != nil {
		t.Fatalf("DisparityBounds raw: %v", err)
	}
	if rawDb.Unit != NumberUnit {
		t.Errorf("raw DisparityBounds unit = %v, want NumberUnit", rawDb.Unit)
	}

	// center/spread Sample bounds propagate the sample's own unit.
	cb, err := sx.CenterBounds(misrate)
	if err != nil {
		t.Fatalf("Sample.CenterBounds: %v", err)
	}
	if cb.Unit != testSecond {
		t.Errorf("Sample CenterBounds unit = %v, want testSecond", cb.Unit)
	}

	spb, err := sx.SpreadBounds(misrate)
	if err != nil {
		t.Fatalf("Sample.SpreadBounds: %v", err)
	}
	if spb.Unit != testSecond {
		t.Errorf("Sample SpreadBounds unit = %v, want testSecond", spb.Unit)
	}

	// shift Sample bounds propagate finer(x, y); identical units -> that unit.
	shb, err := sx.ShiftBounds(sy, misrate)
	if err != nil {
		t.Fatalf("Sample.ShiftBounds: %v", err)
	}
	if shb.Unit != testSecond {
		t.Errorf("Sample ShiftBounds unit = %v, want testSecond", shb.Unit)
	}
}

// --- raw-bounds misrate domain --------------------------------------------
//
// The raw bounds API (native slice, double misrate) must reject an out-of-[0,1]
// or NaN misrate with the domain/misrate AssumptionError. This covers a
// one-sample (CenterBounds) and a two-sample (ShiftBounds) entry. Go has no
// typed Probability wrapper: the Sample methods take the same plain float64
// misrate and delegate to this raw validation, so the raw path is the single
// domain gate.
func TestRawBoundsRejectMisrateDomain(t *testing.T) {
	x := []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	y := []float64{2, 4, 6, 8, 10, 12, 14, 16, 18, 20}

	badMisrates := []float64{2.0, -0.1, math.NaN()}

	assertDomainMisrate := func(t *testing.T, err error, label string) {
		t.Helper()
		if err == nil {
			t.Fatalf("%s: expected domain/misrate error, got nil", label)
			return
		}
		ae, ok := err.(*AssumptionError)
		if !ok {
			t.Fatalf("%s: expected *AssumptionError, got %T: %v", label, err, err)
			return
		}
		if ae.Violation.ID != Domain {
			t.Errorf("%s: id = %q, want %q", label, ae.Violation.ID, Domain)
		}
		if ae.Violation.Subject != SubjectMisrate {
			t.Errorf("%s: subject = %q, want %q", label, ae.Violation.Subject, SubjectMisrate)
		}
	}

	for _, m := range badMisrates {
		m := m
		t.Run("CenterBounds", func(t *testing.T) {
			_, err := CenterBounds(x, m, false)
			assertDomainMisrate(t, err, "CenterBounds")
		})
		t.Run("ShiftBounds", func(t *testing.T) {
			_, err := ShiftBounds(x, y, m, false)
			assertDomainMisrate(t, err, "ShiftBounds")
		})
	}
}

// --- n==2 center midpoint order-symmetry -----------------------------------
//
// Forward guard: pins EXACT order-symmetry of the n==2 center midpoint
// (0.5*a + 0.5*b). assumeSorted=true is REQUIRED so the midpoint sees the raw
// element order: the normalizing sort would otherwise canonicalize the order
// and hide an order-dependent formula. An order-dependent midpoint such as
// a+(b-a)*0.5 rounds differently per order (-3.4 forward but
// -3.4000000000000004 reversed for this input); the exact bit-equality
// assertions below fail on any such replacement.
func TestCenterMidpointOrderSymmetry(t *testing.T) {
	forward, err := Center([]float64{-5.0, -1.8}, true)
	if err != nil {
		t.Fatalf("Center forward: %v", err)
	}
	reverse, err := Center([]float64{-1.8, -5.0}, true)
	if err != nil {
		t.Fatalf("Center reverse: %v", err)
	}
	if forward != reverse {
		t.Errorf("center order asymmetry: forward=%v (bits %#x) reverse=%v (bits %#x)",
			forward, math.Float64bits(forward), reverse, math.Float64bits(reverse))
	}
	if forward != -3.4 {
		t.Errorf("center([-5.0,-1.8]) = %v, want exactly -3.4", forward)
	}
}
